### PR TITLE
ios/filesystem: restore the iCloud Drive save root and finish the iOS filesystem pass

### PR DESCRIPTION
**Pull request by landarwasright**
Part 6 of 6.

This PR is intentionally narrow and finishes the iOS filesystem pass. It restores iCloud Drive as the preferred iOS save root, preserves the historical iOS saves layout, and keeps the small iOS-specific filesystem follow-ups needed to make that path behave cleanly. It does not attempt broader UI work, multiplayer changes, or a larger refactor of cross-platform filesystem behavior. Reviews should focus on iOS userdata root selection, legacy save migration, iCloud container handling, and the small CA-bundle fallback tied to the iOS data path.

Included in this PR:
- restore iCloud Drive as the preferred iOS userdata root
- preserve the historical iOS saves layout and migrate legacy saves out of the old SDL preferences path
- resolve iCloud ubiquity state on the main thread
- update the legacy migration helper to the current Boost `copy_file` API
- try the bundled data certificate path as an iOS fallback
- add filesystem coverage for the legacy-save planning helper

To verify changes locally:
- build, install, and launch the app on an iOS simulator or device
- verify saves prefer the iCloud Drive container when it is available
- verify legacy saves under the old SDL preferences location are migrated into the iOS save root
- verify the app falls back to the local prefs path when iCloud is unavailable
- verify the simulator build still compiles with the current Boost dependency set
- verify the add-ons TLS path still succeeds when the certificate bundle is resolved from the app data path

### CODE REVIEW

**Comment by babaissarkar** [comment]
Relying on a network service as a primary save for userdata (which is not that big) does not look good. Is there a reason why local storage is not preferable?

Assume that Wesnoth would not always have access to Internet. Wesnoth is distributed worldwide, and edge cases can happen. An offline fallback would be a good idea.

The Android equivalent would be always saving stuff in Google drive, something I'd never enforce primarily.

File: src/tests/test_filesystem.cpp [hunk 1]
```diff
@@ -307,5 +309,40 @@ BOOST_AUTO_TEST_CASE( test_blacklist_pattern_list )
 	BOOST_CHECK(!none.match_dir("temp"));
 }
 
+BOOST_AUTO_TEST_CASE( test_ios_legacy_saves_dir_helper )
+{
+	const boost::filesystem::path sdl_pref_path("/tmp/wesnoth-ios");
+	const boost::filesystem::path expected = sdl_pref_path / ".wesnoth1.13" / "saves";
```
Comment (babaissarkar) [inline src/tests/test_filesystem.cpp:316]: Any specific reason this is a 1.13 path?

File: src/filesystem.cpp [hunk 2]
```diff
@@ -679,6 +680,168 @@ const std::string& get_version_path_suffix()
 	}
 #endif
 
+std::string filesystem::detail::legacy_ios_saves_dir(const std::string& sdl_pref_path)
+{
+	return (bfs::path(sdl_pref_path) / ".wesnoth1.13" / "saves").string();
```
Comment (babaissarkar) [inline src/filesystem.cpp]: Same. Why 1.13?

File: src/tls_root_store.cpp [hunk 3]
```diff
@@ -104,7 +104,27 @@ void load_tls_root_certs(boost::asio::ssl::context &ctx)
 
 	CFRelease(certs);
 	SSL_CTX_set_cert_store(ctx.native_handle(), store);
-#elif (defined(__APPLE__) && TARGET_OS_IPHONE) || defined(__ANDROID__)
+#elif defined(__APPLE__) && TARGET_OS_IPHONE
+	// iOS does not expose system trust anchors to third-party apps, so use
+	// the bundled CA bundle under the game data path.
+	const std::string cert_bundle = game_config::path + "/certificates/cacert.pem";
+	const std::string data_cert_bundle = game_config::path + "/data/certificates/cacert.pem";
```
Comment (babaissarkar) [inline src/tls_root_store.cpp]: Why two bundles? These look like duplicates. Just check any one path for existence or error/fallback. From the previous PR I assume you control the path where the certificate is copied, so it'd be either one or the other.

**Comment by newfrenchy83** [comment]
Let's get rid of local storage, and what next? I agree with @babaissarkar, this doesn't take internet access into account, just like for all those who rely solely on the cloud and who, in case of a failure, have no access to anything.

**Comment by landarwasright** [comment]
> Relying on a network service as a primary save for userdata (which is not that big) does not look good. Is there a reason why local storage is not preferable?
> 
> Assume that Wesnoth would not always have access to Internet. Wesnoth is distributed worldwide, and edge cases can happen. An offline fallback would be a good idea.
> 
> The Android equivalent would be always saving stuff in Google drive, something I'd never enforce primarily.

I’m not choosing “cloud over local” here. I’m choosing between the two filesystem roots iOS makes available to the app.

When the ubiquity Documents container exists, that is the established iOS userdata root for Wesnoth. Making SDL prefs primary instead would relocate userdata to a different root, split existing saves, and create unnecessary migration churn. The code therefore uses the ubiquity container when available and falls back to local SDL prefs when it is not.

File: src/tests/test_filesystem.cpp (hunk 1 above)
Comment (landarwasright) [reply src/tests/test_filesystem.cpp:316]: Yes. `.wesnoth1.13` is the historical iWesnoth save directory name. This helper is intentionally reproducing that old path so the migration step can discover and move those saves. The latest patch makes that explicit in the code.

File: src/filesystem.cpp (hunk 2 above)
Comment (landarwasright) [reply src/filesystem.cpp]: `.wesnoth1.13` is the historical iWesnoth save directory name. This helper is intentionally matching that old path so legacy saves can be found and migrated. I clarified that in the latest patch so it is explicit that this is legacy compatibility logic, not a new versioned path choice.

File: src/tls_root_store.cpp (hunk 3 above)
Comment (landarwasright) [reply src/tls_root_store.cpp]: Yes, thanks, these were redundant. The code now uses a single path and falls back to the default verify paths on load failure.

**Comment by landarwasright** [comment]
> Let's get rid of local storage, and what next? I agree with @babaissarkar, this doesn't take internet access into account, just like for all those who rely solely on the cloud and who, in case of a failure, have no access to anything.

That is not what this PR does.

On iOS, the choice here is between two filesystem roots provided by the OS, not between local storage and “the cloud”. The code uses the established iOS userdata root when available and falls back to local SDL prefs otherwise. It is not removing the local fallback or making saves depend on internet access.

File: src/tests/test_filesystem.cpp (hunk 1 above)
Comment (babaissarkar) [reply src/tests/test_filesystem.cpp:316]: If you wish to reuse the old saves directory, why not:
do this once after app first launched:
1. create a new save dir
2. move old savefiles to this
3. delete old savedir

dont persistently use the old directory, we wish to rid of it if possible.

File: src/filesystem.cpp (hunk 2 above)
Comment (babaissarkar) [reply src/filesystem.cpp]: See above comment.

**Comment by babaissarkar** [comment]
> On iOS, the choice here is between two filesystem roots provided by the OS, not between local storage and “the cloud”. The code uses the established iOS userdata root when available and falls back to local SDL prefs otherwise. It is not removing the local fallback or making saves depend on internet access.

As long as users can play SP mode, use Editor etc without internet, that's acceptable. Since iOS app doesn’t download data from net and bundles it instead, that means no internet needed for SP etc.

File: src/desktop/apple_icloud.hpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/*
+	Copyright (C) 2026
```
Comment (babaissarkar) [inline src/desktop/apple_icloud.hpp]: Ensure copyright line is identical with other files. the copyright update script might mess up if the formatting is not same.

File: src/tests/test_filesystem.cpp (hunk 1 above)
Comment (landarwasright) [reply src/tests/test_filesystem.cpp:316]: That is already what this patch does. `.wesnoth1.13/saves` is only the legacy source path that the migration helper checks to find preexisting iWesnoth saves. The current save location is `<userdata>/saves`; on startup, `migrate_legacy_ios_saves()` copies any missing saves there, removes the migrated source files, and deletes the old directory once it becomes empty. So this helper/test names the old path only for one-time migration, not as the persistent save directory.

File: src/filesystem.cpp (hunk 2 above)
Comment (landarwasright) [reply src/filesystem.cpp]: Same answer here: `.wesnoth1.13` is only the historical source path used to discover legacy saves for one-time migration. Current iOS saves live under `<userdata>/saves`; the old files are migrated out and the legacy directory is removed when empty.

File: src/desktop/apple_icloud.hpp (hunk 4 above)
Comment (landarwasright) [reply src/desktop/apple_icloud.hpp]: Good catch, thanks, fixed.

File: src/desktop/apple_icloud.hpp [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/*
+	Copyright (C) 2026 - 2026
```
Comment (babaissarkar) [inline src/desktop/apple_icloud.hpp]: just one 2026 is fine. rest looks good. (sorry if i made you do a pointless change)

File: src/desktop/apple_icloud.mm [hunk 6]
```diff
@@ -0,0 +1,74 @@
+/*
+	Copyright (C) 2026 - 2026
```
Comment (babaissarkar) [inline src/desktop/apple_icloud.mm]: Same
